src/graph/impl/KokkosGraph_Distance2MIS_impl.hpp

```diff
@@ -0,0 +1,737 @@
+  //   (higher priority = less preferred to being in the independent set)
+
+  static constexpr status_t IN_SET = 0;
+  static constexpr status_t OUT_SET = ~IN_SET;
+
+  D2_MIS_Luby(const rowmap_t& rowmap_, const entries_t& entries_)
+    : rowmap(rowmap_), entries(entries_), numVerts(rowmap.extent(0) - 1)
+  {
+    status_t i = numVerts + 1;
+    nvBits = 0;
+    while(i)
+    {
+      i >>= 1;
+      nvBits++;
+    }
+    //Each value in rowStatus represents the status and priority of each row.
+    //Each value in colStatus represents the lowest nonzero priority of any row adjacent to the column.
+    //  This counts up monotonically as vertices are eliminated (given status OUT_SET)
+    rowStatus = status_view_t(Kokkos::ViewAllocateWithoutInitializing("RowStatus"), numVerts);
+    colStatus = status_view_t(Kokkos::ViewAllocateWithoutInitializing("ColStatus"), numVerts);
+    allWorklists = Kokkos::View<lno_t**, Kokkos::LayoutLeft, mem_space>(Kokkos::ViewAllocateWithoutInitializing("AllWorklists"), numVerts, 3);
+  }
+
+  struct RefreshRowStatus
+  {
+    RefreshRowStatus(const status_view_t& rowStatus_, const lno_view_t& worklist_, lno_t nvBits_, int round)
+      : rowStatus(rowStatus_), worklist(worklist_), nvBits(nvBits_)
+    {
+      hashedRound = xorshiftHash(round);
+    }
+
+    KOKKOS_INLINE_FUNCTION void operator()(lno_t w) const
+    {
+      lno_t i = worklist(w);
+      //Combine vertex and round to get some pseudorandom priority bits that change each round
+      status_t priority = xorshiftHash(i + hashedRound);
+      //Generate unique status per row, with IN_SET < status < OUT_SET,
+      int priorityBits = sizeof(status_t) * 8 - nvBits;
+      status_t priorityMask = 1;
+      priorityMask <<= priorityBits;
+      priorityMask--;
+      status_t newStatus = (status_t) (i + 1) + ((priority & priorityMask) << nvBits);
+      if(newStatus == OUT_SET)
+        newStatus--;
+      rowStatus(i) = newStatus;
+    }
+
+    status_view_t rowStatus;
+    lno_view_t worklist;
+    int nvBits;
+    uint32_t hashedRound;
+  };
+
+  struct RefreshColStatus
+  {
+    RefreshColStatus(const status_view_t& colStatus_, const lno_view_t& worklist_, const status_view_t& rowStatus_, const rowmap_t& rowmap_, const entries_t& entries_, lno_t nv_)
+      : colStatus(colStatus_), worklist(worklist_), rowStatus(rowStatus_), rowmap(rowmap_), entries(entries_), nv(nv_)
+    {}
+
+    KOKKOS_INLINE_FUNCTION void operator()(lno_t w) const
+    {
+      lno_t i = worklist(w);
+      //iterate over {i} union the neighbors of i, to find
+      //minimum status.
+      status_t s = OUT_SET;
+      size_type rowBegin = rowmap(i);
+      size_type rowEnd = rowmap(i + 1);
+      for(size_type j = rowBegin; j <= rowEnd; j++)
+      {
+        lno_t nei = (j == rowEnd) ? i : entries(j);
+        if(nei <= nv)
+        {
+          status_t neiStat = rowStatus(nei);
+          if(neiStat < s)
+            s = neiStat;
+        }
+      }
+      if(s == IN_SET)
```

Comment:
Doesn't this need a team barrier here ? s could be set by some thread right ?

Reply:
same comment applies in other places below but this may not matter just making sure.

Reply:
Everything is RangePolicy currently, so s is local to a single CUDA thread.